### PR TITLE
libtomcrypt: fixed dependencies for asymmetric cryptography

### PR DESCRIPTION
**Pull request by lelegard**
In HomeBrew, libtomcrypt had all asymmetric cryptography disabled because no bignum mathematical library was set. Using RSA, ECC and the like was impossible with libtomcrypt in a HomeBrew environment. On other systems, it works.

Required fixes:
- Add the required compilation options to enable LTM and GMP descriptors.
- Build the test program on a separate command because of additional mathematical library requirement.
- Add dependencies on libtommath and gmp, the two supported bignum mathematical libraries.

Note: On Ubuntu, for instance, libtomcrypt depends on libtommath and gmp. With HomeBrew, the two dependencies were missing.

$ apt depends libtomcrypt1
libtomcrypt1
  Depends: libc6 (>= 2.17)
  Depends: libgmp10 (>= 2:6.2.1+dfsg1)
  Depends: libtommath1 (>= 1.1.0)

- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

### CODE REVIEW

**Comment by lelegard** [comment]
Hi HomeBrew maintainers,

Could you please help on this one. The CI failed on macOS11-arm64 with the following message:
~~~
==> Testing libtomcrypt
==> ./test
dyld: Library not loaded: @loader_path/../../../../../opt/libtommath/lib/libtommath.1.dylib
    Referenced from: /private/tmp/libtomcrypt-test-20230824-29211-1af4hzi/./test
    Reason: image not found
~~~

During libtomcrypt test phase, the library `libtommath.1.dylib` is not found while libtommath was added in the dependencies. The build phase for the library and the test program was correct, including any reference to `libtommath.1.dylib`.

On my macOS13-arm64, the command `brew test` went fine with the updated formula, after `brew install --from-source`.

Is there any specific setup to resolve the dependencies during the test phase in the HomeBrew CI environment?

Thanks for your assistance.

File: Formula/lib/libtomcrypt.rb
Comment (ZhongRuoyu) [inline Formula/lib/libtomcrypt.rb:30]: Did you forget to add `-lgmp`? Also, let's build shared libraries too. That will also need `depends_on "libtool" => :build`.

```suggestion
    ENV.append_to_cflags "-DUSE_LTM -DLTM_DESC -DGMP_DESC"
    ENV.append "EXTRALIBS", "-lgmp -ltommath"
    system "make", "-f", "makefile.shared"
    system "make", "-f", "makefile.shared", "test"
    system "./test"
    system "make", "-f", "makefile.shared", "install", "PREFIX=#{prefix}"
```

**Review by ZhongRuoyu — CHANGES_REQUESTED** [review]
In `homebrew/core`, bottles are built with relative install names, so it is generally not safe to move executables to new locations. You should be able to reproduce the failure locally (on macOS) by setting `HOMEBREW_RELOCATABLE_INSTALL_NAMES`.

Test executables should, generally, be built during the test stage. This makes sure the library installed can be used. However, considering that this is a crypto library, it makes sense to run tests during the build stage. So I suggest resolving that by moving the current test to the `install` method.

That said, please also add another new test that builds against the library, to make sure it can be used. Please see [`libxcrypt.rb`](https://github.com/Homebrew/homebrew-core/blob/1aa36b1d93b4ee968d8d355640735f5ec21e7262/Formula/lib/libxcrypt.rb#L33-L62) for an example.

File: Formula/lib/libtomcrypt.rb
Comment (ZhongRuoyu) [reply Formula/lib/libtomcrypt.rb:30]: See also: https://github.com/Homebrew/homebrew-core/pull/22218.

Comment (lelegard) [reply Formula/lib/libtomcrypt.rb:30]: Hi @ZhongRuoyu 

Thanks a lot for your help but I am not sure to understand all your proposed changes. I am not familiar enough with all the Ruby code in formulae.

The original formula was not written by me. I just wanted to reenable asymmetric crypto in libtomcrypt (without it, it broke the port of some test applications). I didn't want to interfere with the structure of the existing formula and I tried to find the minimal required changes.

I have briefly read the discussion in #22218. It is weird that the build with libtommath and gmp was made optional and finally completely removed. This makes libtomcrypt useless in asymmetric crypto. It needs at least one MP math backend.

I also read the discussion whether gmp should be used or not. I think that libtomcrypt should be built with the two. This is the choice of the application to select the proper MP backend at run time. Traditionally, libtomcrypt was used with libtommath (same author) but, nowadays, gmp is much faster than libtommath, twice as fast in average according to my recent tests. This is why the two must be referenced during the build.

Concerning your proposed changes, you run the test application as part of the build. Why not, but would this fix the issue in the test part of the formula? I assume that this test part is the one that is run by `brew test`, which would still fail in the CI then, isn't it?

Comment (ZhongRuoyu) [reply Formula/lib/libtomcrypt.rb:30]: > I just wanted to reenable asymmetric crypto in libtomcrypt (without it, it broke the port of some test applications). I didn't want to interfere with the structure of the existing formula and I tried to find the minimal required changes.

Yes, that's understandable. What I am proposing, on top of your changes, are: 1. to add shared libraries because [it is Homebrew's policy](https://docs.brew.sh/Acceptable-Formulae#static-libraries) that shared libraries should be preferred; 2. to run tests during the build because that works around the "Library not loaded" issue, and more importantly, verify that the crypto library behaves as expected.

> It is weird that the build with libtommath and gmp was made optional and finally completely removed.

> I also read the discussion whether gmp should be used or not.

Yes, that's because `homebrew/core` no longer allows optional dependencies. We can re-enable them like what you're proposing here, though.

> but would this fix the issue in the test part of the formula?

Yes, it would. The direct reason why you encountered the "Library not loaded" failure was the moving of the test executable, from `pkgshare` to the test path. (Your changes were not the root cause.)

Comment (lelegard) [reply Formula/lib/libtomcrypt.rb:30]: > However, a consequence of 2. is that we need a new test that builds a simple program against the built library. Are you willing to do that? If not, I can come up with one.

I will do it. I have plenty of libtomcrypt apps and I can extract a small one.

Comment (ZhongRuoyu) [reply Formula/lib/libtomcrypt.rb:30]: Thanks! Sorry for the language; `s/Are you willing to/Would you like to/`.

Comment (lelegard) [reply Formula/lib/libtomcrypt.rb]: > Sorry for the language

Don't worry, I am not a native English either.
I have pushed an update. It builds and passes the test locally. Let's see the CI...

Comment (lelegard) [reply Formula/lib/libtomcrypt.rb]: All CI checks passed this time. Thanks @ZhongRuoyu for your help. The PR just needs your approval now.

**Review by ZhongRuoyu — DISMISSED** [review]
Looks good. Thank you @lelegard!

**Pull request closed without merging**
